### ISSUE TITLE
dor-utils virtual-merge script broken (blocked by dor-services #651)

### ISSUE THREAD

**Issue by blalbrit** [sul-dlss-deprecated/dor-utils#17]
### Note to devs:

Scroll down to [Aug 6, 2019 comments](https://github.com/sul-dlss/dor-utils/issues/17#issuecomment-518753815) for more current and dev oriented information.

<hr />

Running into problems with virtual-merge since most recent patch party updates:

method for invoking virtual merge to-date has been:
`rvm use 2.2.2 do bundle exec bin/virtual-merge --environment=production` (followed by a bunch of druids)

This produces the error message:
Could not find activemodel-3.2.22.4 in any of the sources
Run `bundle install` to install missing gems.

Attempts to `bundle install` fail.

Attempting to run virtual-merge with the default rvm version of Ruby (2.2.4) produce:
bundler: failed to load command: bin/virtual-merge (bin/virtual-merge)
NameError: uninitialized constant VirtualMergeTool::Honeybadger
  bin/virtual-merge:47:in `rescue in block in run’
  bin/virtual-merge:43:in `block in run’
  bin/virtual-merge:42:in `each’
  bin/virtual-merge:42:in `run’
  bin/virtual-merge:113:in `<top (required)>'

This is a production service, so would be great to get it back up and running as soon as possible (with appropriate updates to make it less fragile?).

**Comment by blalbrit** [comment]
@tingulfsen and @cbeer - any thoughts on this ^^ ? possibly related to https://github.com/sul-dlss/dor-utils/pull/16 ?

**Comment by blalbrit** [comment]
virtual merge no longer works as expected at all:

```
bundler: failed to load command: bin/virtual-merge (bin/virtual-merge)
NoMethodError: undefined method `url' for nil:NilClass
Did you mean?  URI
  /home/lyberadmin/dor-utils/releases/20181212211828/config/environments/production.rb:23:in `block (2 levels) in <top (required)>'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `instance_eval'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `eval_or_yield'
  /usr/local/rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/delegate.rb:341:in `block in delegating_block'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/confstruct-0.2.7/lib/confstruct/hash_with_struct_access.rb:195:in `method_missing'
  /home/lyberadmin/dor-utils/releases/20181212211828/config/environments/production.rb:22:in `block in <top (required)>'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `instance_eval'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `eval_or_yield'
  /usr/local/rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/delegate.rb:341:in `block in delegating_block'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/confstruct-0.2.7/lib/confstruct/configuration.rb:21:in `configure'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/dor-services-6.0.1/lib/dor/config.rb:26:in `block in configure'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/activesupport-5.2.2/lib/active_support/callbacks.rb:132:in `run_callbacks'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.3.0/gems/dor-services-6.0.1/lib/dor/config.rb:25:in `configure'
  /home/lyberadmin/dor-utils/releases/20181212211828/config/environments/production.rb:3:in `<top (required)>'
  bin/virtual-merge:111:in `require_relative'
  bin/virtual-merge:111:in `<top (required)>'
```

I have gone to running from an earlier deploy in order to manage production requests.

cc. @vivnwong

**Comment by blalbrit** [comment]
There are no longer earlier deploys for this, still seeing

```
[lyberadmin@sul-lyberservices-prod current]$ bundle exec bin/virtual-merge --purge --environment=production  druid:ds623xk2719 druid:cf088nn0221 druid:cw049rd7067 druid:rz597dq8213 druid:dg148hz5693 druid:qf251yh6066
bundler: failed to load command: bin/virtual-merge (bin/virtual-merge)
NoMethodError: undefined method `url' for nil:NilClass
  /home/lyberadmin/dor-utils/releases/20190506164505/config/environments/production.rb:23:in `block (2 levels) in <top (required)>'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `instance_eval'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `eval_or_yield'
  /usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/delegate.rb:349:in `block in delegating_block'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/confstruct-0.2.7/lib/confstruct/hash_with_struct_access.rb:195:in `method_missing'
  /home/lyberadmin/dor-utils/releases/20190506164505/config/environments/production.rb:22:in `block in <top (required)>'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `instance_eval'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/confstruct-0.2.7/lib/confstruct/utils.rb:5:in `eval_or_yield'
  /usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/delegate.rb:349:in `block in delegating_block'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/confstruct-0.2.7/lib/confstruct/configuration.rb:21:in `configure'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/config.rb:27:in `block in configure'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:132:in `run_callbacks'
  /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/config.rb:26:in `configure'
  /home/lyberadmin/dor-utils/releases/20190506164505/config/environments/production.rb:3:in `<top (required)>'
  bin/virtual-merge:111:in `require_relative'
  bin/virtual-merge:111:in `<top (required)>'
```

I consider this now a blocker for production requests. @vivnwong - I'll send out a service announcement and direct stakeholders to you for further communication.

**Comment by justinlittman** [comment]
Made the necessary config change in stage and production. @blalbrit can you test at your convenience?

**Comment by hannahfrost** [comment]
Need to confirm the fix with testing.

**Comment by hannahfrost** [comment]
This was tested and remains broken. Let's leave the ticket open until we address the larger virtual-merge issue in the upcoming workcycle.

**Comment by ndushay** [comment]
tried it again:

```
[lyberadmin@sul-lyberservices-prod current]$ bundle exec bin/virtual-merge --purge --environment=production  druid:ds623xk2719 druid:cf088nn0221 druid:cw049rd7067 druid:rz597dq8213 druid:dg148hz5693 druid:qf251yh6066
DEPRECATION WARNING: Dor::Config -- solrizer configuration is deprecated. Please use solr instead. (called from <top (required)> at /home/lyberadmin/dor-utils/releases/20190805145213/config/environments/production.rb:3)
bundler: failed to load command: bin/virtual-merge (bin/virtual-merge)
NameError: uninitialized constant VirtualMergeTool::Honeybadger
  bin/virtual-merge:47:in `rescue in block in run'
  bin/virtual-merge:43:in `block in run'
  bin/virtual-merge:42:in `each'
  bin/virtual-merge:42:in `run'
  bin/virtual-merge:113:in `<top (required)>'
```

Will see if I can fix configs

**Comment by jcoyne** [comment]
@ndushay https://github.com/sul-dlss/dor-utils/pull/64

**Comment by ndushay** [comment]
just deployed with PR #63 in;  trying again -- it SEEMS happy:

```
[lyberadmin@sul-lyberservices-prod current]$ bundle exec bin/virtual-merge --purge --e=production --debug  druid:ds623xk2719 druid:cf088nn0221 druid:cw049rd7067 druid:rz597dq8213 druid:dg148hz5693 druid:qf251yh6066
DEPRECATION WARNING: Dor::Config -- solrizer configuration is deprecated. Please use solr instead. (called from <top (required)> at /home/lyberadmin/dor-utils/releases/20190806004452/config/environments/production.rb:3)
```

@jcoyne is getting rid of the deprecation warning super simple?

**Comment by ndushay** [comment]
@blalbrit @andrewjbtw @hannahfrost - this seems to be fixed now???   Can you please confirm?

**Comment by jcoyne** [comment]
@ndushay I don't know. That file isn't managed by git.

**Comment by ndushay** [comment]
@jcoyne is it something that can be changed in shared_configs?  Or is dor-utils using an old way to connect with Solr?

**Comment by jcoyne** [comment]
It's not using shared_configs

**Comment by jcoyne** [comment]
I think we just want to change this line:
```
45:  solrizer.url 'http://sul-solr.stanford.edu/solr/argo3_prod'
```
to 
```
45:  solr.url 'http://sul-solr.stanford.edu/solr/argo3_prod'
```

**Comment by blalbrit** [comment]
@ndushay - thanks for doing some preliminary testing. The combination you point to in the objects above failed. Here's what is currently wrong:
* contentMD for ds623xk2719 was not built (note that the datastream is empty - should have resources for each "child" object): 
![Screenshot 2019-08-06 09 48 34](https://user-images.githubusercontent.com/796990/62559321-98521280-b82f-11e9-8daf-cfd949046d6a.png)
* qf251yh6066 does not have isConstituentOf (as you pointed out). 

This leads me to believe the process failed before completing updating the list of child druids and writing the contentMD to the parent object (ds623xk2719).

**Comment by ndushay** [comment]
The script `bin/virtual-merge` is broken.  Here's what the script is supposed to do:

Given a list of druids:
- the first druid is assumed to be the virtual collection object; subsequent druids are child objects
- the virtual collection object should get `contentMetadata` built with resources from/for each child object
  e.g.

- each child object should have "isConstituentOf" relationship to virtual collection object added to `RELS-EXT`.

```xml
<rdf:RDF xmlns:fedora="info:fedora/fedora-system:def/relations-external#" xmlns:fedora-model="info:fedora/fedora-system:def/model#" xmlns:hydra="http://projecthydra.org/ns/relations#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
  <rdf:Description rdf:about="info:fedora/druid:cf088nn0221">
    <hydra:isGovernedBy rdf:resource="info:fedora/druid:xs835jp8197"/>
    <fedora-model:hasModel rdf:resource="info:fedora/afmodel:Dor_Abstract"/>
    <fedora-model:hasModel rdf:resource="info:fedora/afmodel:Dor_Item"/>
    <fedora:isConstituentOf rdf:resource="info:fedora/druid:ds623xk2719"/>
  </rdf:Description>
</rdf:RDF>
```



#### Running the script on stage data
(it is idempotent with --purge, so it is okay to re-run it)
- on `lyberservices-test`:
  - `bundle exec bin/virtual-merge --purge --e=production  druid:ds623xk2719 druid:cf088nn0221

#### Running the script on production data
(it is idempotent with --purge, so it is okay to re-run it)
- bundle exec bin/virtual-merge --purge --e=production --debug  druid:ds623xk2719 druid:cf088nn0221 druid:cw049rd7067 druid:rz597dq8213 druid:dg148hz5693 druid:qf251yh6066

**Comment by ndushay** [comment]
Errors from `/home/lyberadmin/dor-utils/shared/log/virtual-merge.log`:

```
I, [2019-08-06T10:27:54.007363 #7332]  INFO -- : merging child druid:cf088nn0221 into parent druid:ds623xk2719
I, [2019-08-06T10:27:54.025346 #7332]  INFO -- : adding virtual resource cf088nn0221_1 of type image to parent druid:ds623xk2719
E, [2019-08-06T10:27:54.028088 #7332] ERROR -- : undefined method `+' for nil:NilClass: /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/datastreams/content_metadata_ds.rb:171:in `add_virtual_resource'
bin/virtual-merge:34:in `block in merge'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:238:in `block in each'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `upto'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `each'
bin/virtual-merge:32:in `merge'
bin/virtual-merge:44:in `block in run'
bin/virtual-merge:42:in `each'
bin/virtual-merge:42:in `run'
bin/virtual-merge:114:in `<top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `load'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `<main>'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `<main>'
I, [2019-08-06T10:27:54.248218 #7332]  INFO -- : merging child druid:cw049rd7067 into parent druid:ds623xk2719
I, [2019-08-06T10:27:54.265812 #7332]  INFO -- : adding virtual resource cw049rd7067_1 of type image to parent druid:ds623xk2719
E, [2019-08-06T10:27:54.266273 #7332] ERROR -- : undefined method `+' for nil:NilClass: /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/datastreams/content_metadata_ds.rb:171:in `add_virtual_resource'
bin/virtual-merge:34:in `block in merge'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:238:in `block in each'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `upto'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `each'
bin/virtual-merge:32:in `merge'
bin/virtual-merge:44:in `block in run'
bin/virtual-merge:42:in `each'
bin/virtual-merge:42:in `run'
bin/virtual-merge:114:in `<top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `load'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `<main>'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `<main>'
I, [2019-08-06T10:27:54.669978 #7332]  INFO -- : merging child druid:rz597dq8213 into parent druid:ds623xk2719
I, [2019-08-06T10:27:54.708369 #7332]  INFO -- : adding virtual resource rz597dq8213_1 of type image to parent druid:ds623xk2719
E, [2019-08-06T10:27:54.708854 #7332] ERROR -- : undefined method `+' for nil:NilClass: /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/datastreams/content_metadata_ds.rb:171:in `add_virtual_resource'
bin/virtual-merge:34:in `block in merge'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:238:in `block in each'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `upto'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `each'
bin/virtual-merge:32:in `merge'
bin/virtual-merge:44:in `block in run'
bin/virtual-merge:42:in `each'
bin/virtual-merge:42:in `run'
bin/virtual-merge:114:in `<top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `load'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `<main>'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `<main>'
I, [2019-08-06T10:27:55.090960 #7332]  INFO -- : merging child druid:dg148hz5693 into parent druid:ds623xk2719
I, [2019-08-06T10:27:55.129132 #7332]  INFO -- : adding virtual resource dg148hz5693_1 of type image to parent druid:ds623xk2719
E, [2019-08-06T10:27:55.129626 #7332] ERROR -- : undefined method `+' for nil:NilClass: /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/datastreams/content_metadata_ds.rb:171:in `add_virtual_resource'
bin/virtual-merge:34:in `block in merge'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:238:in `block in each'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `upto'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `each'
bin/virtual-merge:32:in `merge'
bin/virtual-merge:44:in `block in run'
bin/virtual-merge:42:in `each'
bin/virtual-merge:42:in `run'
bin/virtual-merge:114:in `<top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `load'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `<main>'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `<main>'
I, [2019-08-06T10:27:55.326064 #7332]  INFO -- : merging child druid:qf251yh6066 into parent druid:ds623xk2719
I, [2019-08-06T10:27:55.344362 #7332]  INFO -- : adding virtual resource qf251yh6066_1 of type image to parent druid:ds623xk2719
E, [2019-08-06T10:27:55.344825 #7332] ERROR -- : undefined method `+' for nil:NilClass: /home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/dor-services-6.8.0/lib/dor/datastreams/content_metadata_ds.rb:171:in `add_virtual_resource'
bin/virtual-merge:34:in `block in merge'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:238:in `block in each'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `upto'
/home/lyberadmin/dor-utils/shared/bundle/ruby/2.5.0/gems/nokogiri-1.10.3/lib/nokogiri/xml/node_set.rb:237:in `each'
bin/virtual-merge:32:in `merge'
bin/virtual-merge:44:in `block in run'
bin/virtual-merge:42:in `each'
bin/virtual-merge:42:in `run'
bin/virtual-merge:114:in `<top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
/usr/local/rvm/rubies/ruby-2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.5.3@global/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `load'
/usr/local/rvm/gems/ruby-2.5.3/bin/bundle:23:in `<main>'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.3/bin/ruby_executable_hooks:24:in `<main>'
```

**Comment by ndushay** [comment]
Note that it's not clear how well this script was working before 2017:

every add in the production log had this ...

```
[lyberadmin@sul-lyberservices-prod log]$ head -100 virtual-merge.log 
I, [2016-03-14T21:48:15.128955 #4321]  INFO -- : merging child druid:by372cw6149 into parent druid:cr768rj4665
I, [2016-03-14T21:48:15.271602 #4321]  INFO -- : adding virtual resource by372cw6149_1 of type image to parent druid:cr768rj4665
E, [2016-03-14T21:48:17.764443 #4321] ERROR -- : undefined method `[]' for nil:NilClass: /home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/rsolr-ext-1.0.3/lib/rsolr-ext/response/docs.rb:4:in `extended'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/rsolr-ext-1.0.3/lib/rsolr-ext/response.rb:17:in `extend'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/rsolr-ext-1.0.3/lib/rsolr-ext/response.rb:17:in `initialize'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/rsolr-ext-1.0.3/lib/rsolr-ext/client.rb:24:in `new'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/rsolr-ext-1.0.3/lib/rsolr-ext/client.rb:24:in `find'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/services/search_service.rb:79:in `query'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor-services.rb:45:in `find_all'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/models/workflow_object.rb:15:in `find_by_name'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/workflow/document.rb:45:in `definition'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/workflow/document.rb:72:in `processes'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/workflow/document.rb:97:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/datastreams/workflow_ds.rb:84:in `block in to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/datastreams/workflow_ds.rb:84:in `each'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/datastreams/workflow_ds.rb:84:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/active-fedora-5.7.1/lib/active_fedora/base.rb:314:in `block in to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/active-fedora-5.7.1/lib/active_fedora/base.rb:313:in `each_value'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/active-fedora-5.7.1/lib/active_fedora/base.rb:313:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/models/identifiable.rb:60:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/models/processable.rb:153:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/models/describable.rb:189:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/models/discoverable.rb:7:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/dor-services-4.24.0/lib/dor/models/releaseable.rb:343:in `to_solr'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/active-fedora-5.7.1/lib/active_fedora/persistence.rb:89:in `update_index'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/active-fedora-5.7.1/lib/active_fedora/persistence.rb:14:in `save'
/home/lyberadmin/dor-utils/shared/bundle/ruby/1.9.1/gems/active-fedora-5.7.1/lib/active_fedora/validations.rb:50:in `save'
bin/virtual-merge:37:in `merge'
bin/virtual-merge:44:in `block in run'
bin/virtual-merge:42:in `each'
bin/virtual-merge:42:in `run'
bin/virtual-merge:111:in `<main>'
I```

**Comment by ndushay** [comment]
After PR sul-dlss/dor-services#654 is merged:

- [ ] cut a new 7.2.x release of dor-services gem
- [ ] have dor-utils use new 7.2.x version of dor-services gem (a new PR)
- [ ] fix config on lyberservices-test for dor-utils gem
- [ ] deploy to stage;  test

- [ ] fix config on lyberservices-prod for dor-utils gem
- [ ] deploy to prod;  test

**Comment by ndushay** [comment]
To test without purge, Ben ran this set of druids:

```
druid:zf679by5082   druid:rd568jh8421   druid:qn433mv8404   druid:vv514yj6391   druid:nc748qq3109   druid:cf321hq1109   druid:rv744bs9996   druid:nd248vc4577   druid:mw296vz8416   druid:yv857sp4788   druid:nb609ch5458   druid:bq517vv4529   druid:bc590sx0783   druid:dy321ft2845   druid:kx465kj7218   druid:sg443fc3130   druid:bn544pw4982   druid:yy229rm7942   druid:cw914sp1793   druid:nc885fz1469   druid:xy319gd1135   druid:cs336pp5461   druid:bc548zm4977   druid:pm261wb2761   druid:mv656kt1053   druid:fq465pf2622   druid:qz871rf0800   druid:jc999bt9479   druid:bq142hs3019   druid:xy644nb7147   druid:ph443dp7186   druid:gf471wc5247   druid:sz724sh2978   druid:bs455gf5196   druid:xj092nw5863   druid:sr637jm4870   druid:xz942cs6171   druid:sf317jr6869   druid:vy641xb1727   druid:nz977jz3441   druid:sy603cq5237   druid:kq619bn7165   druid:nf356bt9973   druid:jm145sv7913   druid:kw807nn8490   druid:gw157np8532   druid:kp536dn3168   druid:xp129yr0598   druid:pn205dn5965   druid:ht814fb8641   druid:gk741xs6180   druid:js886gr3660   druid:gc846dh7586   druid:vn602ny2408   druid:tp494tg7603   druid:cr783fd5418   druid:qw297rk9647   druid:fj902gs8671   druid:qy104bp8765   druid:wj814rg3269   druid:gm018zb5003   druid:mm600zy6999   druid:jw504gc4763   druid:xx979ng0368   druid:bf053dt5519   druid:ct224nj4308   druid:wm917gp1754   druid:rx153hy8863   druid:wg050pj9541   druid:xj993hm6242   druid:yk644jk7842   druid:hb266xh9324   druid:hd338vh7227   druid:tx673sq5521   druid:ny707km1539   druid:hm029yj8026   druid:gk984pq3125   druid:mj643pt4063   druid:cb223wy1402   druid:qb078tr0659   druid:gm225nw5930   druid:yn709gr5473   druid:sg056pf5604
```

and all was well for parent except resource 21:

![image](https://user-images.githubusercontent.com/96775/62804324-c76fba80-baa1-11e9-9baa-f4b6b68ef77a.png)

and looking at resource 21:
the mimetype of the first resource is suspect:
![image](https://user-images.githubusercontent.com/96775/62804359-dfdfd500-baa1-11e9-88cf-cdec94c0d247.png)

Ben says the problem is:
> it’s the thumbs.db that shouldn’t have been in there anyway!
> so - code is working properly. data from content owners is… as expected :wink:

**Comment by ndushay** [comment]
confirmed with Ben;  merged to master and deployed to stage and prod.
